File: swift/Tests/FundamentalVanillaTests/RestingMotionTests.swift
import Foundation
import Testing
#if canImport(simd)
import simd
#endif
@testable import FundamentalVanilla
import FundamentalCore

// The resting-motion floor (board #24 "resting liveliness") — the Swift mirror of
// packages/core/src/resting-motion-floor.test.ts. DECLARED, default OFF: an idle field with the ambient
// drift stilled settles to rest (the default step is unchanged — the golden stays byte-identical); the
// `.thermal` and `.flow` floors each keep it alive with nothing drawn; `.flow` is divergence-free by
// construction; `.thermal` draws through the seeded rng so a run reproduces; reduced motion contributes nothing.
@Suite("Resting-motion floor")
struct RestingMotionTests {

    private func idle(seed: UInt32, _ resting: RestingMotion?, host: HeadlessFieldHost = HeadlessFieldHost()) -> (FieldField, HeadlessFieldHost) {
        let field = FieldField(host: host, options: .init(ambientOrbit: 0, ambientWander: 0, restingMotion: resting, rng: seededRng(seed)))
        return (field, host)
    }

    private func positions(_ field: FieldField) -> [Float] {
        var out = [Float](repeating: 0, count: field.particleCount() * 5)
        let n = field.readParticles(into: &out)
        var pos: [Float] = []
        pos.reserveCapacity(n * 2)
        for i in 0..<n { pos.append(out[5 * i]); pos.append(out[5 * i + 1]) }
        return pos
    }

    /// mean per-frame displacement over `frames` frames (edge wraps excluded).
    private func meanStep(_ field: FieldField, _ host: HeadlessFieldHost, from frame: Int, frames: Int = 30) -> Float {
        var sum: Float = 0; var n = 0
        var prev = positions(field)
        for f in 0..<frames {
            host.fire(at: TimeInterval(frame + f) / 60)
            let cur = positions(field)
            var i = 0
            while i + 1 < cur.count {
                let dx = cur[i] - prev[i], dy = cur[i + 1] - prev[i + 1]
                if abs(dx) <= 50, abs(dy) <= 50 { sum += (dx * dx + dy * dy).squareRoot(); n += 1 }
                i += 2
            }
            prev = cur
        }
        return n == 0 ? 0 : sum / Float(n)
    }

    private func settled(_ resting: RestingMotion?, seed: UInt32 = 1) -> Float {
        let (field, host) = idle(seed: seed, resting)
        for i in 0..<240 { host.fire(at: TimeInterval(i) / 60) } // let the seeded launch velocities decay
        let s = meanStep(field, host, from: 240)
        field.destroy()
        return s
    }

    @Test("restingFlow is divergence-free by construction, and not trivial")
    func divergenceFree() {
        let rng = seededRng(42)
        let h: Float = 1e-2
        var maxDiv: Float = 0, maxMag: Float = 0
        for _ in 0..<200 {
            let x = rng() * 800, y = rng() * 600, phase = rng() * 2 * .pi
            let dvxdx = (restingFlow(x: x + h, y: y, phase: phase).x - restingFlow(x: x - h, y: y, phase: phase).x) / (2 * h)
            let dvydy = (restingFlow(x: x, y: y + h, phase: phase).y - restingFlow(x: x, y: y - h, phase: phase).y) / (2 * h)
            maxDiv = max(maxDiv, abs(dvxdx + dvydy))
            let v = restingFlow(x: x, y: y, phase: phase)
            maxMag = max(maxMag, (v.x * v.x + v.y * v.y).squareRoot())
        }
        #expect(maxDiv < 1e-3) // f32 finite differences
        #expect(maxMag > 0.5)
    }

    @Test("OFF is really off: an idle field with the ambient drift stilled settles to rest")
    func offIsOff() {
        #expect(settled(nil) < 0.02)
    }

    @Test("the thermal floor keeps an idle field alive with nothing drawn")
    func thermalFloor() {
        #expect(settled(RestingMotion(mode: .thermal)) > 0.08)
    }

    @Test("the flow floor keeps an idle field alive with nothing drawn")
    func flowFloor() {
        #expect(settled(RestingMotion(mode: .flow)) > 0.08)
    }

    @Test("strength 0 is the off path; strength scales the floor")
    func strengthScales() {
        #expect(settled(RestingMotion(mode: .thermal, strength: 0)) < 0.02)
        let one = settled(RestingMotion(mode: .thermal, strength: 1))
        let two = settled(RestingMotion(mode: .thermal, strength: 2))
        #expect(two > one * 1.5)
    }

    @Test("reduced motion: the floor contributes exactly nothing (dt = 0)")
    func reducedMotion() {
        for mode in [RestingMotionMode.thermal, .flow] {
            let host = HeadlessFieldHost()
            host.prefersReducedMotion = true
            let (field, _) = idle(seed: 1, RestingMotion(mode: mode), host: host)
            let before = positions(field)
            for i in 0..<60 { host.fire(at: TimeInterval(i) / 60) }
            #expect(positions(field) == before)
            field.destroy()
        }
    }

    @Test("thermal draws through the injected rng: a seeded run reproduces, a different seed diverges")
    func seededDeterminism() {
        func run(_ seed: UInt32) -> [Float] {
            let (field, host) = idle(seed: seed, RestingMotion(mode: .thermal))
            for i in 0..<120 { host.fire(at: TimeInterval(i) / 60) }
            let p = positions(field)
            field.destroy()
            return p
        }
        #expect(run(7) == run(7))
        #expect(run(7) != run(8))
    }
}

File: android/fundamental-core/src/test/kotlin/com/fundamental/core/RestingMotionTests.kt
package com.fundamental.core

import com.fundamental.core.engine.RESTING_FLOW_K
import com.fundamental.core.engine.RestingMotion
import com.fundamental.core.engine.RestingMotionMode
import com.fundamental.core.engine.restingFlow
import com.fundamental.core.runtime.FieldController
import kotlin.math.abs
import kotlin.math.hypot
import kotlin.random.Random
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNotEquals
import kotlin.test.assertTrue

// The resting-motion floor (board #24 "resting liveliness") — the Kotlin mirror of
// packages/core/src/resting-motion-floor.test.ts. DECLARED, default OFF: an idle field with the ambient
// drift stilled settles to rest (the default step is unchanged — the golden stays byte-identical); the
// THERMAL and FLOW floors each keep it alive with nothing drawn; FLOW is divergence-free by construction;
// THERMAL draws through the seeded rng so a run reproduces.
class RestingMotionTests {

    private fun idle(seed: Long, resting: RestingMotion?): FieldController =
        FieldController(800f, 600f, 0f, particleCount = 200, seed = seed, ambientOrbit = 0f, ambientWander = 0f, restingMotion = resting)
            .also { it.setFormation("ambient") }

    private fun positions(c: FieldController): List<Pair<Float, Float>> = c.particles.map { it.position.x to it.position.y }

    /** mean per-frame displacement over `frames` frames (edge wraps excluded). */
    private fun meanStep(c: FieldController, frames: Int = 30): Float {
        var sum = 0f; var n = 0
        var prev = positions(c)
        repeat(frames) {
            c.tick()
            val cur = positions(c)
            for (i in cur.indices) {
                val dx = cur[i].first - prev[i].first; val dy = cur[i].second - prev[i].second
                if (abs(dx) > 50f || abs(dy) > 50f) continue
                sum += hypot(dx, dy); n++
            }
            prev = cur
        }
        return if (n == 0) 0f else sum / n
    }

    private fun settled(resting: RestingMotion?, seed: Long = 1): Float {
        val c = idle(seed, resting)
        repeat(240) { c.tick() } // let the seeded launch velocities decay (0.95^240 ≈ 4e-6)
        return meanStep(c)
    }

    @Test
    fun restingFlowIsDivergenceFreeByConstructionAndNotTrivial() {
        val rng = Random(42); val h = 1e-2f
        var maxDiv = 0f; var maxMag = 0f
        repeat(200) {
            val x = rng.nextFloat() * 800f; val y = rng.nextFloat() * 600f; val phase = rng.nextFloat() * 6.2832f
            val dvxdx = (restingFlow(x + h, y, phase).x - restingFlow(x - h, y, phase).x) / (2f * h)
            val dvydy = (restingFlow(x, y + h, phase).y - restingFlow(x, y - h, phase).y) / (2f * h)
            maxDiv = maxOf(maxDiv, abs(dvxdx + dvydy))
            val v = restingFlow(x, y, phase); maxMag = maxOf(maxMag, hypot(v.x, v.y))
        }
        assertTrue(maxDiv < 1e-3f, "∂vx/∂x + ∂vy/∂y should vanish everywhere (max $maxDiv; f32 finite differences)")
        assertTrue(maxMag > 0.5f, "the flow field is not the zero field")
        assertTrue(RESTING_FLOW_K > 0f)
    }

    @Test
    fun offIsReallyOffAnIdleFieldSettlesToRest() {
        val s = settled(null)
        assertTrue(s < 0.02f, "idle mean step should be ≈0 with the floor off, got $s")
    }

    @Test
    fun theThermalFloorKeepsAnIdleFieldAlive() {
        val s = settled(RestingMotion(RestingMotionMode.THERMAL))
        assertTrue(s > 0.08f, "thermal floor should lift the idle mean step well above rest, got $s")
    }

    @Test
    fun theFlowFloorKeepsAnIdleFieldAlive() {
        val s = settled(RestingMotion(RestingMotionMode.FLOW))
        assertTrue(s > 0.08f, "flow floor should lift the idle mean step well above rest, got $s")
    }

    @Test
    fun strengthZeroIsTheOffPathAndStrengthScalesTheFloor() {
        assertTrue(settled(RestingMotion(RestingMotionMode.THERMAL, 0f)) < 0.02f, "strength 0 is off")
        val one = settled(RestingMotion(RestingMotionMode.THERMAL, 1f)); val two = settled(RestingMotion(RestingMotionMode.THERMAL, 2f))
        assertTrue(two > one * 1.5f, "strength 2 moves matter more than strength 1 ($two vs $one)")
    }

    @Test
    fun thermalDrawsThroughTheSeededRngSoARunReproduces() {
        fun run(seed: Long): List<Pair<Float, Float>> { val c = idle(seed, RestingMotion(RestingMotionMode.THERMAL)); repeat(120) { c.tick() }; return positions(c) }
        assertEquals(run(7), run(7), "same seed → identical positions")
        assertNotEquals(run(7), run(8), "different seed → different positions")
    }
}
